br/pkg/storage/mapstore.go

```diff
@@ -0,0 +1,304 @@
+	_, ok := s.dataStore[name]
+	return ok, nil
+}
+
+// Open a Reader by file path.
+func (s *mapStorage) Open(ctx context.Context, filePath string) (ExternalFileReader, error) {
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	default:
+		// continue on
+	}
+	if !path.IsAbs(filePath) {
+		return nil, errors.Errorf("file name is not an absolute path: %s", filePath)
+	}
+	s.RLock()
+	defer s.RUnlock()
+	theFile, ok := s.dataStore[filePath]
+	if !ok {
+		return nil, errors.Errorf("cannot find the file: %s", filePath)
+	}
+	theFile.RLock()
+	defer theFile.RUnlock()
+	r := bytes.NewReader(theFile.Data)
+	return &mapFileReader{
+		Reader: r,
+	}, nil
+}
+
+// WalkDir traverse all the files in a dir.
+func (s *mapStorage) WalkDir(ctx context.Context, opt *WalkOption, fn func(string, int64) error) error {
+	s.RLock()
+	defer s.RUnlock()
+	for fileName, file := range s.dataStore {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+			// continue on
+		}
+		if opt != nil {
+			if len(opt.SubDir) > 0 {
+				if !strings.HasPrefix(fileName, opt.SubDir) {
+					continue
+				}
+			}
+			if len(opt.ObjPrefix) > 0 {
+				baseName := path.Base(fileName)
+				if !strings.HasPrefix(baseName, opt.ObjPrefix) {
+					continue
+				}
+			}
+		}
+		file.RLock()
+		fileSize := len(file.Data)
+		file.RUnlock()
+		if err := fn(fileName, int64(fileSize)); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (s *mapStorage) URI() string {
+	return "mapstore://"
+}
+
+// Create implements ExternalStorage interface.
+func (s *mapStorage) Create(ctx context.Context, name string) (ExternalFileWriter, error) {
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	default:
+		// continue on
+	}
+	if !path.IsAbs(name) {
+		return nil, errors.Errorf("file name is not an absolute path: %s", name)
+	}
+	s.Lock()
+	defer s.Unlock()
+	if _, ok := s.dataStore[name]; ok {
+		return nil, errors.Errorf("the file already exists: %s", name)
+	}
+	theFile := &mapFile{}
+	s.dataStore[name] = theFile
+	return &mapFileWriter{
+		file: theFile,
+	}, nil
+}
+
+// Rename implements ExternalStorage interface.
+func (s *mapStorage) Rename(ctx context.Context, oldFileName, newFileName string) error {
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	default:
+		// continue on
+	}
+	if !path.IsAbs(newFileName) {
+		return errors.Errorf("new file name is not an absolute path: %s", newFileName)
+	}
+	s.Lock()
+	defer s.Unlock()
+	theFile, ok := s.dataStore[oldFileName]
+	if !ok {
+		return errors.Errorf("the file doesn't exist: %s", oldFileName)
+	}
+	if _, ok := s.dataStore[newFileName]; ok {
+		return errors.Errorf("the new file has already existed: %s", newFileName)
+	}
+	s.dataStore[newFileName] = theFile
+	delete(s.dataStore, oldFileName)
+	return nil
+}
+
+type mapFileReader struct {
+	*bytes.Reader
+	isClosed atomic.Bool
+}
+
+func (r *mapFileReader) Read(p []byte) (int, error) {
+	if r.isClosed.Load() {
+		return -1, errors.New("reader closed")
```

Comment:
It seems we should return `0, io.EOF` instead. Refer https://pkg.go.dev/io#Reader.